src/events/index.ts

```diff
@@ -0,0 +1,75 @@
+import Emittery from "emittery"
+
+import { CID } from "../common/cid.js"
+import { DistinctivePath, Partition, Partitioned } from "../path/index.js"
+import { Ucan } from "../ucan/types.js"
+
+export type Emitter<EventMap> = InstanceType<typeof Emittery<EventMap, EventMap>>
+export type Listener<EventMap> = (eventData: EventMap) => void | Promise<void>
+
+export { Emitter as EventEmitter, Listener as EventListener }
+
+/** @protected */
+export { Emittery as EmitterClass }
+
+/**
+ * Events interface.
+ *
+ * Subscribe to events using `on` and unsubscribe using `off`.
+ * There's also `once`, `onAny`, `offAny`, `anyEvent` and `events`.
+ *
+ * ```ts
+ * fileSystem.on("local-change", ({ path, root }) => {
+ *   console.log("The file system has changed locally 🔔")
+ *   console.log("Changed path:", path)
+ *   console.log("New data root CID:", root)
+ * })
+ *
+ * fileSystem.off("publish")
+ * ```
+ */
+export type ListenTo<EventMap> = Pick<
+  Emitter<EventMap>,
+  "on" | "onAny" | "off" | "offAny" | "once" | "anyEvent" | "events"
+>
+
+export type AuthorityRequestor = {
+  "challenge": undefined // TODO
+}
+
+export type AuthorityProvider = {
+  "approved": undefined
+  "challenge": undefined // TODO
+  "dismissed": undefined
+  "query": Record<string, any> // TODO
+}
+
+export type FileSystem = {
+  "local-change": { dataRoot: CID; path: DistinctivePath<Partitioned<Partition>> }
+  "publish": { dataRoot: CID; proofs: Ucan[] }
+}
```

Comment:
It looks like these events may not be narrowing to `local-change` and `publish`. With the following code:

```typescript
  fs.on('local-change', event => {
    event
  })
```

The type of `event` is:

```typescript
(parameter) event: {
    dataRoot: odd.CID<unknown, number, number, Version>;
    path: odd.path.DistinctivePath<[odd.path.Partition, ...string[]]>;
} | {
    dataRoot: odd.CID<unknown, number, number, Version>;
    proofs: odd.ucan.Ucan[];
}
```

which includes the parameters for both `local-only` and `publish`, but we should only get the parameters for `local-change`.

Reply:
Oh good catch! Fixed 👍